Add properties object to Grants & update hashing algorithm (breaking) (#25)
Logius-standaarden/fsc-core: specifications.md

The data in Contracts is hashed and signed. To ensure hash consistency, the FSC specification defines the exact order in which fields should be hashed. However, the `properties` object can contain arbitrary data types, making it impossible to predetermine a fixed field ordering as is done for Grants.
To address this issue, the `properties` object must be converted to a canonical JSON representation as specified in [[RFC8785]]. This canonicalization ensures deterministic hashing regardless of the original field ordering or data structure variations.

#### requirements

- Each Grant **MAY** contain a `properties` object. 
- The `properties` object **MUST** have the key `properties`.
- The `properties` object **MUST** be a valid JSON object that can contain any number of key-value pairs.
- When provided, the `properties` object **MUST** be included in the access token as specified in the [access token section].(#access_token)
- When provided, the `properties` object **MUST** be included in the Grant Hash as specified in the [Grant hash section].(#grant-hash)

#### Security Considerations

> - The size of the `properties` object should be limited to prevent excessive data transfer and storage. It is recommended to implement a size limit of 1MB for the serialized properties object.

Reviewer: ```suggestion
- It is **RECOMMENDED** to implement a size limit of 1MB for the serialized properties object. This limit should prevent excessive data transfer and storage.
```

1. Convert the value of `contract.content.algorithm` to an int32 and surround it with dollar signs (`$`). When using the `SHA3-512` algorithm this would result in `$1$`. 
   To convert the hash algorithm to an integer, see the [type mapping](#type_mapping_hash_algorithm)
1. Add `1$` as suffix to the string created in step 13. This is the enum `HASH_TYPE_CONTRACT` as defined in the field `.components.schemas.HashType` of the [OpenAPI Specification](manager.yaml) as int32. If the string created in step 13 is `$1$`, the result should now be `$1$1$`
1. Add the Base64 generated in step 4 as suffix to the string generated in step 5.

### Grant hash {#grant_hash}

The Grant hash is used in the access token request to identify the Grant which contains the authorization for the connection to the Service. 
The `iv` (Initialization vector) field is included in the Grant hash to create a Grant hash that references to a single Contract. 
The Grant hash can be created by executing the following steps:

1. Add the fields `iv` and `group_id` to the `data` object of the Grant. 
1. Set `contract.content.group_id` as value for `data.group_id`.
1. Set `contract.content.iv` as value for `data.iv`.

> 1. Convert the Grant to Canonical JSON data as described in [[RFC8785]]

Reviewer: Let's be explicit here if `{ Data: { ... }}` should be hashed, or, only the contents of `Data`.

The `cnf.x5t#S256` claim **MUST** contain the certificate thumbprint of the X.509 certificate provided by the client requesting the token according to [section 3.1] of [[RFC8705]].
The `act` claim **MUST** be set when an access token is generated for a Peer who is connecting to the Service on behalf of another Peer. I.e. the authorization to connect has been granted using a DelegatedServiceConnectionGrant.
The `pdi` claim **MUST** be set when an access token is generated for a Service which is being offered on behalf of another Peer. 

> The `prp` claim **MUST** be set when the Grant contains a `properties` object in the field `grant.data.properties`. 

Reviewer: ```suggestion
The `prp` claim **MUST** be set when the Grant contains a the `properties` object for the field `grant.data.properties`. 
```

Author: I Used a different sentence structure.
